File: 01_GOF23/DP_examples/CSharp/01_Creational/Factory/Product.cs
namespace Creational.Factory;

public interface IShape {
    void Draw();
    void Fill();
}

public enum ShapeKind {
    None, Circle, Square, Rectangle, BigCircle
}
internal class Circle : IShape {
    public void Draw() => Console.WriteLine("Draw a Circle");
    public void Fill() => Console.WriteLine("Fill red");
}
internal class Square : IShape {
    public void Draw() => Console.WriteLine("Draw a Square");
    public void Fill() => Console.WriteLine("Fill Green");
}
internal class Rectangle : IShape {
    public void Draw() => Console.WriteLine("Draw a Rectangle");
    public void Fill() => Console.WriteLine("Fill Blue");
}

internal class UnknownShape : IShape {
    public void Draw() { }
    public void Fill() { }
}

File: 01_GOF23/DP_examples/Rust/01_Creational/Factory/product.rs
pub trait IShape {
    fn draw(&self);
    fn fill(&self);
}

#[derive(Eq, Hash, PartialEq, Debug)]
pub enum ShapeKind {
    Unknown,
    Circle,
    Square,
    Rectangle,
    BigCircle,
}

pub struct Circle;

impl IShape for Circle {
    fn draw(&self) {
        println!("Draw a Circle");
    }
    
    fn fill(&self) {
        println!("Fill red");
    }
}

pub struct Square;

impl IShape for Square {
    fn draw(&self) {
        println!("Draw a Square");
    }
    
    fn fill(&self) {
        println!("Fill Green");
    }
}

pub struct Rectangle;

impl IShape for Rectangle {
    fn draw(&self) {
        println!("Draw a Rectangle");
    }
    
    fn fill(&self) {
        println!("Fill Blue");
    }
}

pub struct UnknownShape;

impl IShape for UnknownShape {
    fn draw(&self) {}
    fn fill(&self) {}
}